### ISSUE TITLE
Torch compile with deepspeed zero optimizer bug

### ISSUE THREAD

**Issue by hust-nj** [pytorch/pytorch#96960]
### 🐛 Describe the bug

I create a model by
```
model, optimizer, _, _ = deepspeed.initialize(args=args, model=model, model_parameters=param_groups, dist_init_required=False)
model = torch.compile(model)
```
However, it raises the following error:
```
The above exception was the direct cause of the following exception:

Traceback (most recent call last):
  File "/tmp/amlt_code/code/examples/crab_clip/train.py", line 189, in <module>
    main()
  File "/tmp/amlt_code/code/examples/crab_clip/train.py", line 185, in main
    runner.run(val_loader=val_loader)
  File "/tmp/amlt_code/code/utils/runner.py", line 149, in run
    self.train(train_loader, val_loader)
  File "/tmp/amlt_code/code/utils/runner.py", line 184, in train
    loss = self._training_callback(
  File "/tmp/amlt_code/code/examples/crab_clip/train.py", line 49, in clip_trainig_batch
    loss = model(**batch)
  File "/opt/conda/lib/python3.10/site-packages/torch/_dynamo/eval_frame.py", line 118, in __call__
    return self.dynamo_ctx(self._orig_mod.__call__)(*args, **kwargs)
  File "/opt/conda/lib/python3.10/site-packages/torch/_dynamo/eval_frame.py", line 254, in _fn
    return fn(*args, **kwargs)
  File "/opt/conda/lib/python3.10/site-packages/torch/nn/modules/module.py", line 1533, in _call_impl
    return forward_call(*args, **kwargs)
  File "/home/xxx/.local/lib/python3.10/site-packages/deepspeed/utils/nvtx.py", line 10, in wrapped_fn
    get_accelerator().range_push(func.__qualname__)
  File "/home/xxx/.local/lib/python3.10/site-packages/deepspeed/utils/nvtx.py", line 11, in <resume in wrapped_fn>
    ret_val = func(*args, **kwargs)
  File "/opt/conda/lib/python3.10/site-packages/torch/_dynamo/eval_frame.py", line 391, in catch_errors
    return callback(frame, cache_size, hooks)
  File "/opt/conda/lib/python3.10/site-packages/torch/_dynamo/convert_frame.py", line 406, in _convert_frame
    result = inner_convert(frame, cache_size, hooks)
  File "/opt/conda/lib/python3.10/site-packages/torch/_dynamo/convert_frame.py", line 105, in _fn
    return fn(*args, **kwargs)
  File "/opt/conda/lib/python3.10/site-packages/torch/_dynamo/convert_frame.py", line 263, in _convert_frame_assert
    return _compile(
  File "/opt/conda/lib/python3.10/site-packages/torch/_dynamo/utils.py", line 164, in time_wrapper
    r = func(*args, **kwargs)
  File "/opt/conda/lib/python3.10/site-packages/torch/_dynamo/convert_frame.py", line 396, in _compile
    raise InternalTorchDynamoError() from e
torch._dynamo.exc.InternalTorchDynamoError
Traceback (most recent call last):
  File "/opt/conda/lib/python3.10/site-packages/torch/_dynamo/convert_frame.py", line 366, in _compile
    check_fn = CheckFunctionManager(
  File "/opt/conda/lib/python3.10/site-packages/torch/_dynamo/guards.py", line 640, in __init__
    guard.create(local_builder, global_builder)
  File "/opt/conda/lib/python3.10/site-packages/torch/_guards.py", line 159, in create
    return self.create_fn(self.source.select(local_builder, global_builder), self)
  File "/opt/conda/lib/python3.10/site-packages/torch/_dynamo/guards.py", line 200, in ID_MATCH
    self._produce_guard_code(guard, [code])
  File "/opt/conda/lib/python3.10/site-packages/torch/_dynamo/guards.py", line 563, in _produce_guard_code
    obj_ref = weakref.ref(guarded_object)
TypeError: cannot create weak reference to 'ZeroStageEnum' object


You can suppress this exception and fall back to eager by setting:
    torch._dynamo.config.suppress_errors = True
```

### Versions

Collecting environment information...
PyTorch version: 2.1.0.dev20230309+cu118
Is debug build: False
CUDA used to build PyTorch: 11.8
ROCM used to build PyTorch: N/A

OS: Ubuntu 20.04.5 LTS (x86_64)
GCC version: (Ubuntu 9.4.0-1ubuntu1~20.04.1) 9.4.0
Clang version: Could not collect
CMake version: version 3.25.0
Libc version: glibc-2.31

Python version: 3.10.9 (main, Jan 11 2023, 15:21:40) [GCC 11.2.0] (64-bit runtime)
Python platform: Linux-4.15.0-197-generic-x86_64-with-glibc2.31
Is CUDA available: True
CUDA runtime version: 11.8.89
CUDA_MODULE_LOADING set to: LAZY
GPU models and configuration:
GPU 0: Tesla V100-SXM2-32GB
GPU 1: Tesla V100-SXM2-32GB
GPU 2: Tesla V100-SXM2-32GB
GPU 3: Tesla V100-SXM2-32GB

Nvidia driver version: 470.141.03
cuDNN version: Probably one of the following:
/usr/lib/x86_64-linux-gnu/libcudnn.so.8.7.0
/usr/lib/x86_64-linux-gnu/libcudnn_adv_infer.so.8.7.0
/usr/lib/x86_64-linux-gnu/libcudnn_adv_train.so.8.7.0
/usr/lib/x86_64-linux-gnu/libcudnn_cnn_infer.so.8.7.0
/usr/lib/x86_64-linux-gnu/libcudnn_cnn_train.so.8.7.0
/usr/lib/x86_64-linux-gnu/libcudnn_ops_infer.so.8.7.0
/usr/lib/x86_64-linux-gnu/libcudnn_ops_train.so.8.7.0
HIP runtime version: N/A
MIOpen runtime version: N/A
Is XNNPACK available: True

CPU:
Architecture:                    x86_64
CPU op-mode(s):                  32-bit, 64-bit
Byte Order:                      Little Endian
Address sizes:                   46 bits physical, 48 bits virtual
CPU(s):                          96
On-line CPU(s) list:             0-95
Thread(s) per core:              2
Core(s) per socket:              24
Socket(s):                       2
NUMA node(s):                    2
Vendor ID:                       GenuineIntel
CPU family:                      6
Model:                           85
Model name:                      Intel(R) Xeon(R) Platinum 8160 CPU @ 2.10GHz
Stepping:                        4
CPU MHz:                         2966.386
BogoMIPS:                        4200.00
Virtualization:                  VT-x
L1d cache:                       1.5 MiB
L1i cache:                       1.5 MiB
L2 cache:                        48 MiB
L3 cache:                        66 MiB
NUMA node0 CPU(s):               0,2,4,6,8,10,12,14,16,18,20,22,24,26,28,30,32,34,36,38,40,42,44,46,48,50,52,54,56,58,60,62,64,66,68,70,72,74,76,78,80,82,84,86,88,90,92,94
NUMA node1 CPU(s):               1,3,5,7,9,11,13,15,17,19,21,23,25,27,29,31,33,35,37,39,41,43,45,47,49,51,53,55,57,59,61,63,65,67,69,71,73,75,77,79,81,83,85,87,89,91,93,95
Vulnerability Itlb multihit:     KVM: Mitigation: Split huge pages
Vulnerability L1tf:              Mitigation; PTE Inversion; VMX conditional cache flushes, SMT vulnerable
Vulnerability Mds:               Mitigation; Clear CPU buffers; SMT vulnerable
Vulnerability Meltdown:          Mitigation; PTI
Vulnerability Mmio stale data:   Mitigation; Clear CPU buffers; SMT vulnerable
Vulnerability Spec store bypass: Mitigation; Speculative Store Bypass disabled via prctl and seccomp
Vulnerability Spectre v1:        Mitigation; usercopy/swapgs barriers and __user pointer sanitization
Vulnerability Spectre v2:        Mitigation; Retpolines, IBPB conditional, IBRS_FW, STIBP conditional, RSB filling
Vulnerability Srbds:             Not affected
Vulnerability Tsx async abort:   Mitigation; Clear CPU buffers; SMT vulnerable
Flags:                           fpu vme de pse tsc msr pae mce cx8 apic sep mtrr pge mca cmov pat pse36 clflush dts acpi mmx fxsr sse sse2 ss ht tm pbe syscall nx pdpe1gb rdtscp lm constant_tsc art arch_perfmon pebs bts rep_good nopl xtopology nonstop_tsc cpuid aperfmperf pni pclmulqdq dtes64 monitor ds_cpl vmx smx est tm2 ssse3 sdbg fma cx16 xtpr pdcm pcid dca sse4_1 sse4_2 x2apic movbe popcnt tsc_deadline_timer aes xsave avx f16c rdrand lahf_lm abm 3dnowprefetch cpuid_fault epb cat_l3 cdp_l3 invpcid_single pti intel_ppin ssbd mba ibrs ibpb stibp tpr_shadow vnmi flexpriority ept vpid fsgsbase tsc_adjust bmi1 hle avx2 smep bmi2 erms invpcid rtm cqm mpx rdt_a avx512f avx512dq rdseed adx smap clflushopt clwb intel_pt avx512cd avx512bw avx512vl xsaveopt xsavec xgetbv1 xsaves cqm_llc cqm_occup_llc cqm_mbm_total cqm_mbm_local dtherm ida arat pln pts pku ospke md_clear flush_l1d arch_capabilities

Versions of relevant libraries:
[pip3] mypy==1.1.1
[pip3] mypy-extensions==1.0.0
[pip3] numpy==1.23.5
[pip3] pytorch-lightning==2.0.0
[pip3] pytorch-triton==2.0.0+b8b470bc59
[pip3] torch==2.1.0.dev20230309+cu118
[pip3] torchaudio==2.0.0.dev20230309+cu118
[pip3] torchmetrics==0.11.4
[pip3] torchvision==0.15.0.dev20230309+cu118
[conda] mkl                       2023.0.0         h6d00ec8_25399
[conda] mkldnn                    0.16.1                        0    mingfeima
[conda] numpy                     1.23.5                   pypi_0    pypi
[conda] pytorch-triton            2.0.0+b8b470bc59          pypi_0    pypi
[conda] torch                     2.1.0.dev20230309+cu118          pypi_0    pypi
[conda] torchaudio                2.0.0.dev20230309+cu118          pypi_0    pypi
[conda] torchvision               0.15.0.dev20230309+cu118          pypi_0    pypi

cc @ezyang @soumith @msaroufim @wconstab @ngimel @bdhirsh @voznesenskym @penguinwu @anijain2305 @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @Xia-Weiwen @wenzhe-nrv @jiayisunx @desertfire

**Comment by eellison** [comment]
Looks like we are trying to create a weakref to a value type (enum). I think it's an easy fix in this case, although we may run into issues with other classes we can't weakref (for instance, if you use `__slots__` and don't explicitly add a weakref slot)

**Comment by hust-nj** [comment]
> 

Any ideas to resolve this weak ref issue? Thank you.

**Comment by jansel** [comment]
In this case I think we could just catch the `TypeError` and replace the weakref with a strongref.  Perhaps something like:
```
try:
    obj_ref = weakref.ref(guarded_object)
except TypeError:
    obj_ref = lambda: guarded_object
```

This will prevent `guarded_object` from being garbage collected, so maybe we should emit a warning.

**Comment by yanboliang** [comment]
I'll close as it has been fixed by #98652, feel free to reopen if it still there.
